batch_test.go

```diff
@@ -61,22 +58,52 @@ func TestEmptyBatchFails(t *testing.T) {
 	}
 }
 
-func BenchmarkVerifyBatch(b *testing.B) {
-	for _, n := range []int{1, 8, 64, 1024} {
+func BenchmarkBatch(b *testing.B) {
+	for _, n := range []int{1, 8, 64, 1024, 4096, 16384} {
 		b.Run(fmt.Sprint(n), func(b *testing.B) {
+			var msg = []byte("ed25519consensus")
+			b.ResetTimer()
 			b.ReportAllocs()
-			v := NewBatchVerifier()
-			for i := 0; i < n; i++ {
+			for i := 0; i < b.N; i++ {
+				b.StopTimer()
 				pub, priv, _ := ed25519.GenerateKey(nil)
-				msg := []byte("BatchVerifyTest")
-				v.Add(pub, msg, ed25519.Sign(priv, msg))
+				sig := ed25519.Sign(priv, msg)
+				b.StartTimer()
+				v := NewBatchVerifier()
+				for j := 0; j < n; j++ {
+					v.Add(pub, msg, sig)
```

Comment:
Previously, I created a new public key per item. I felt that was a better representation of what is actually going on in a typical batch (it may not affect the current code but could affect future optimization effectiveness).

Reply:
Done! I was a bit worried about the overhead of calling StartTimer/StopTimer in such a hot loop, but the benchmarks look good.

I don't want to pregenerate all the keys and signatures, because assigning them to a slice causes them to escape, making it harder to observe the escape analysis of the arguments of Add.